pubsub/cloud-client/quickstart.py

```diff
@@ -19,76 +19,110 @@
-    print('\nSubscription created: {}\n'.format(subscription.name))
-
-    publish_begin = time.time()
+    try:
+        subscriber.delete_subscription(subscription_path)
+    except NotFound:
+        pass
+    finally:
+        subscription = subscriber.create_subscription(
+            subscription_path, topic_path)
+        print('Subscription created: \"{}\"'.format(subscription.name))
+
+    # `data` is roughly 10 Kb.
+    data = 'x' * 9600
+    # `data` must be a bytestring.
+    data = data.encode('utf-8')
+    # Initialize an empty dictionary to track messages.
+    tracker = dict()
+    pubsub_time = 0.0
 
     # Publish messages.
-    for n in range(num_messages):
-        data = u'Message number {}'.format(n)
-        # Data must be a bytestring
-        data = data.encode('utf-8')
-        # When you publish a message, the client returns a future.
+    for i in range(num_messages):
+        # When we publish a message, the client returns a future.
         future = publisher.publish(topic_path, data=data)
-        print('Published {} of message ID {}.'.format(data, future.result()))
 
-    publish_time = time.time() - publish_begin
+        tracker.update({i: {'index': i,
+                            'pubtime': time.time(),
+                            'subtime': None}})
+
+        # `future.result()` blocks and returns a unique message ID per message.
+        tracker[future.result()] = tracker.pop(i)
 
-    messages = set()
+    print('\nPublished all messages.')
 
     def callback(message):
-        print('Received message: {}'.format(message))
         # Unacknowledged messages will be sent again.
         message.ack()
-        messages.add(message)
-
-    subscribe_begin = time.time()
+        # Update message `subtime`.
+        tracker[message.message_id]['subtime'] = time.time()
 
     # Receive messages. The subscriber is nonblocking.
     subscriber.subscribe(subscription_path, callback=callback)
 
-    print('\nListening for messages on {}...\n'.format(subscription_path))
+    print('\nListening for messages...')
 
     while True:
-        if len(messages) == num_messages:
-            subscribe_time = time.time() - subscribe_begin
-            print("\nReceived all messages.")
-            print("Publish time lapsed: {:.2f}s.".format(publish_time))
-            print("Subscribe time lapsed: {:.2f}s.".format(subscribe_time))
+        # Deplete messages in `tracker` every time we have complete info
+        # of a message.
+        for msg_id in list(tracker):
+            pubtime = tracker[msg_id]['pubtime']
+            subtime = tracker[msg_id]['subtime']
+            if subtime is not None:
+                pubsub_time += subtime - pubtime
+                del tracker[msg_id]
+
+        # Exit if `tracker` is empty i.e. all the messages' publish-subscribe
+        # time have been accounted for.
+        if len(tracker) == 0:
+            print('\nTotal publish to subscribe time for {} messages: \
```

Comment:
I think it would be much more useful to print statistics here.  The total time is mostly determined by the number of messages and batching settings, not system performance.  So I'd recommend making pubsub_time a list of latencies.  And then printing stats of that: average, best time, worst time, and if number of messages > 100, say. 
 10th, 25th ptile, 75th ptile, 90th ptile (if number of record > 100